### PR TITLE
Python APIだけlocal version segmentsを付ける

### LINKED ISSUE #304: release時にpython wheelの名称が被ってアップロードできない
## 不具合の内容

最新版のコアでビルドすると、wheelのアップロードで名前が衝突してアップロードに失敗してエラーになる現象が発生しました。
おそらく、CUDAありと無しでビルドしても同じwhl名になってエラーになっているのかなと思います。

### 現象・ログ

`Upload Python whl to Release`で以下のエラーが出る。

### 再現手順

UPLOAD RELEASEフラグを1にしてビルドする。

### 期待動作

リリースできる

## その他

wheelの名称の最後に`-cuda`なり`-directml`なりを足せると良さそうなのですが、どのあたりを変更すれば良いのかわからずにいます。

### PR DESCRIPTION
**Pull request by qryxip**
## 内容

リリース時にwhlの名前をリネームします。

Maturin側でwhlの名付けをコントロールするのは面倒そうなので、whlの生成後にリネームするようにしました。

## 関連 Issue

Fixes #304.

### CODE REVIEW

**Comment by qryxip** [comment]
例: `voicevox-core-gpu-0.14.0-check.0-cp38-abi3-linux_x86_64.whl`

File: .github/workflows/build_and_deploy.yml [hunk 1]
```diff
@@ -34,41 +34,49 @@ jobs:
...
             features: ""
             target: x86_64-unknown-linux-gnu
             artifact_name: linux-x64-cpu
+            whl_package_name: voicevox-core-cpu
             use_cuda: false
           - os: ubuntu-latest
             features: ""
             target: x86_64-unknown-linux-gnu
             artifact_name: linux-x64-gpu
+            whl_package_name: voicevox-core-gpu
```
Comment (Hiroshiba) [inline .github/workflows/build_and_deploy.yml:67]: windowsのCUDA版と揃えてこうするのはどうでしょう
```suggestion
            whl_package_name: voicevox-core-cuda
```

File: .github/workflows/build_and_deploy.yml [hunk 2]
```diff
@@ -111,7 +119,10 @@ jobs:
         run: |
           pip install -r ./crates/voicevox_core_python_api/requirements.txt
           maturin build --manifest-path ./crates/voicevox_core_python_api/Cargo.toml --features ${{ matrix.features }}, --target ${{ matrix.target }} --release
-          printf '::set-output name=whl::%s\n' "$(find ./target/wheels -type f)"
+          from="$(find ./target/wheels -type f)"
+          to="$(sed 's/voicevox_core/${{ matrix.whl_package_name }}/' <<<"$from")"
+          mv "$from" "$to"
```
Comment (Hiroshiba) [inline .github/workflows/build_and_deploy.yml:124]: 元が`voicevox_core`なので、ひとかたまりな概念はスネークケースを維持しつつ、概念が別れてる部分はケバブケースにする形で`voicevox_core-cuda`にすると本家っぽいかもと思いました！

**Review by Hiroshiba — COMMENTED** [review]
ほぼLGTMの気持ちです！！
命名に関して２点コメントしました！

File: .github/workflows/build_and_deploy.yml (hunk 2 above)
Comment (qryxip) [reply .github/workflows/build_and_deploy.yml:124]: なるほど。そうしました。

File: .github/workflows/build_and_deploy.yml [hunk 3]
```diff
@@ -34,41 +34,49 @@ jobs:
             features: ""
             target: x86_64-pc-windows-msvc
             artifact_name: windows-x64-cpu
+            whl_package_name: voicevox_core-cpu
```
Comment (PickledChair) [inline .github/workflows/build_and_deploy.yml:37]: https://github.com/VOICEVOX/voicevox_core/pull/305#discussion_r990924527 の提案を受け、suffix をケバブケースで結合する形に変更されましたが、 [PEP 427 で定義されている whl のファイル名の形式](https://peps.python.org/pep-0427/#file-name-convention) を参照したところ、ハイフンは区切り文字として特別な意味を持つことがわかりました。

したがってここをケバブケースにすると PEP 427 にそぐわなくなるので、最初のようにアンダースコアで繋ぐ形にするのが良いと思いました。

もう一点、wheel 内の METADATA の方は、パッケージの名前が `voicevox_core` のままであり、wheel のファイル名の先頭に書かれているパッケージ名と不一致であることに問題があるかもしれないことが気になっています（こちらは実際に問題があるかどうかはまだ確認できておりません）。

Comment (qryxip) [reply .github/workflows/build_and_deploy.yml]: > したがってここをケバブケースにすると PEP 427 にそぐわなくなるので、最初のようにアンダースコアで繋ぐ形にするのが良いと思いました。

なるほど。snake_caseにしました。
 
> もう一点、wheel 内の METADATA の方は、パッケージの名前が `voicevox_core` のままであり、wheel のファイル名の先頭に書かれているパッケージ名と不一致であることに問題があるかもしれないことが気になっています（こちらは実際に問題があるかどうかはまだ確認できておりません）。

METADATAのことを失念していました。
パッケージ名についてはCLI引数で設定できないみたいなので、pyproject.tomlを`sed -i`するようにしました。

**Comment by qryxip** [comment]
versionの方に"+cuda"と付けるという手もあるかもしれません。

```toml
[package]
name = "voicevox_core_python_api"
version = "0.14.0-check.0+cuda"
```

この場合Maturinは`voicevox_core-0.14.0_check.0_cuda-cp38-abi3-linux_x86_64.whl`という名前にするみたいです。

**Comment by qryxip** [comment]
パッケージ名を変換しているのはここですね。`+cuda`はPEP440のlocal verisonとして扱われるようです。
https://docs.rs/pep440/latest/pep440/struct.Version.html#method.normalize

whl名として`+`を勝手に`_`に潰していいのかという疑問は残りますが。

**Comment by Hiroshiba** [comment]
どうするのが正解なのか難しいのですが、[pytorchのwheel](https://download.pytorch.org/whl/torch/)に前ならえで行くと、

・バージョンに`+device`にして、
image

・パッケージ名は全部`voicevox_core`にする（pytorchはどれでも「import torch」で利用可能になる）

とかが良いのかなとか思いました！

**Comment by qryxip** [comment]
`+device`には賛成なのですが、pre-release部分が`-check.0`のままでいいのかどうかは議論してもいいのではないかと思いました。whl名がこうなるので。

- `0.14.0-check.0+cuda`
    → `voicevox_core-0.14.0_check.0_cuda-cp38-abi3-linux_x86_64.whl` (Pythonパッケージのバージョンとしておかしいと判断され、フォールバックとして適当にsnake_caseにされる)
- `0.14.0-beta.1+cuda`
    → `voicevox_core-0.14.0b1+cuda-cp38-abi3-linux_x86_64.whl` (PEP440に従いちゃんと正規化される)

いっそのことPython APIだけ`s/check/beta/`するのもありなんじゃないかと思います。

**Comment by qryxip** [comment]
> pytorchはどれでも「import torch」で利用可能になる

本筋からは外れますが、これについてはパッケージ名とモジュール名は独立なのでは?という気がします。
(voicevox_core_python_apiの場合、/crates/voicevox_core_python_api/python/voicevox_coreというディレクトリの名前から取られているはず)

**Comment by Hiroshiba** [comment]
> pre-release部分が-check.0のままでいいのかどうかは議論した方がいいのではないかと思いました。

pre-releaseは適当に作っていてユーザー向けではないということもあって、installできればまあなんでも良いのかなと思っています。
snake_caseに変えてくれるならまあpython wheel任せでも良いのかなーと！

**Comment by qryxip** [comment]
なるほど。ひとまず`+cpu`や`+cuda`をバージョンに付ける方針に切り替えました。

File: .github/workflows/build_and_deploy.yml [hunk 4]
```diff
@@ -104,17 +104,13 @@ jobs:
         shell: bash
         run: |
           for cargo_toml in $( echo ./crates/voicevox_core_*/Cargo.toml ); do
-            set-cargo-version $cargo_toml ${{ env.VERSION }}
+            if [[ "$cargo_toml" =~ voicevox_core_python_api ]]; then
```
Comment (Hiroshiba) [inline .github/workflows/build_and_deploy.yml:107]: 確認なのですが、 `=~` はどういう意味でしょうか 👀

**Review by Hiroshiba — APPROVED** [review]
LGTM！！！

File: .github/workflows/build_and_deploy.yml (hunk 4 above)
Comment (qryxip) [reply .github/workflows/build_and_deploy.yml:107]: `grep -E "$regex" <<<"$text" >/dev/null`に似た動きをするBashの機能であるという認識です。

```console
❯ bash --version
GNU bash, version 5.1.16(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
❯ [[ aaabbbccc =~ bbb ]]; echo "$?"
0
❯ [[ aaabbbccc =~ ddd ]]; echo "$?"
1
❯ [[ aaabbbccc =~ [a-z]+ ]]; echo "$?"
0
❯ [[ aaabbbccc =~ [A-Z]+ ]]; echo "$?"
1
```

(あとGoogle検索は記号の列をちゃんと認識してくれることがあります。`bash =~`で`=~`について検索できると思います)

**Comment by qryxip** [comment]
あと`whl_version_local`は["Local version segments"](https://peps.python.org/pep-0440/#local-version-segments)の意味です。Rustの[`pep440::LocalVersion`](https://docs.rs/pep440/latest/pep440/enum.LocalVersion.html)のように`whl_local_version`とした方がよかったかも。

File: .github/workflows/build_and_deploy.yml (hunk 4 above)
Comment (Hiroshiba) [reply .github/workflows/build_and_deploy.yml:107]: おお、なるほどです！！
そして記号列もちゃんと検索できますね！！盲点でした。

**Review by PickledChair — APPROVED** [review]
LGTM！

https://github.com/VOICEVOX/voicevox_core/pull/305#issuecomment-1276371382 にある PEP440 に従ったバージョン文字列の normalize に伴う問題についても了解しました。

試しに `0.14.0-check.0+cuda` というバージョンを設定した python パッケージをインストールして pyo3 サンプルを実行してみましたが、インストール・動作はとりあえず確認できました。また、パッケージのファイル名におけるバージョン文字列は `0.14.0_check.0_cuda` になりましたが、内部的には以下のように `0.14.0-check.0-cuda` という文字列になっているようでした。

```
$ pip list
Package           Version
----------------- -------------------
numpy             1.23.4
pip               22.3
pydantic          1.10.2
setuptools        65.5.0
typing_extensions 4.4.0
voicevox_core     0.14.0-check.0-cuda
```

（開発バージョンのバージョン文字列を真面目に考えるならもう少し作業が必要ですが、 https://github.com/VOICEVOX/voicevox_core/pull/305#issuecomment-1279992440 のコメントのようにとりあえず release バージョンのことのみ考えるという方針であるということで了解しました。）
